[기본 레이아웃 컴포넌트] 고니(정우시) 미션 제출합니다.

jeongwusi opened the pull request:
안녕하세요 애슐리~!!

리뷰어로 만나뵙게 되어 영광입니다. 잘 부탁드립니다. 😊

# 🎯 1단계. 기본 레이아웃 컴포넌트 ([npm 배포 링크](https://www.npmjs.com/package/goni-layout-component))

## 공통
- [x] 의미있는 커밋 메시지 작성
- [x] PR에 관련 없는 변경 사항이 포함 유무 확인

## Container 컴포넌트
- [x] `minWidth`, `maxWidth` 속성 적용
- [x] npm 배포
- [x] README.md 작성
- [ ] 테스트 코드를 작성하였나요? (선택 사항)

## Grid 컴포넌트
- [x] `rows`, `columns`, `gap` 속성 적용
- [x] npm 배포
- [x] README.md 작성
- [ ] 테스트 코드를 작성하였나요? (선택 사항)

## Flex 컴포넌트
- [x] `direction`, `justify`, `align`, `gap` 속성 적용
- [x] npm 배포
- [x] README.md 작성
- [ ] 테스트 코드를 작성하였나요? (선택 사항)

## 📌 구현한 내용 설명

> React Layout Components 라이브러리를 구현하는 과정에서 다양한 고민과 어려움을 겪었습니다. 아래에서는 라이브러리 개발과정에서의 주요 고민과 해결책, 어려웠던 부분에 대해 설명하겠습니다.

**고민과 어려웠던 점**

컴포넌트 API 설계: 각 레이아웃 컴포넌트의 API를 어떻게 디자인할지 고민이 필요했습니다. 사용자가 쉽게 이해하고 사용할 수 있는 API를 제공하기 위해 다양한 옵션을 고려하였습니다.

CSS 스타일링 관리: 컴포넌트 내부에서 CSS 스타일을 관리하고 적용하는 방법을 고민했습니다. 각 컴포넌트의 스타일을 동적으로 생성하고 적용하는 방법을 선택했습니다.

**해결책과 구현**

컴포넌트 API 설계: 사용자의 입장에서 각 컴포넌트가 어떻게 사용되어야 하는지 고려하여, 필수 및 선택적인 props를 정의했습니다. 또한 TypeScript의 인터페이스를 활용하여 타입 안정성을 보장했습니다.

CSS 스타일링 관리: 컴포넌트 내부에서 CSS 스타일을 동적으로 생성하고 적용하기 위해 React의 style 속성과 CSSProperties를 활용했습니다. 이를 통해 스타일을 컴포넌트와 함께 캡슐화할 수 있었습니다.

**추가적인 고려 사항**

문서화: README.md 파일을 작성하여 라이브러리의 사용 방법과 API를 상세하게 문서화했습니다. 또한 예제 코드와 함께 제공하여 사용자들이 라이브러리를 쉽게 사용할 수 있도록 하였습니다.

패키지 구조: npm 배포를 위해 패키지 구조를 설계하고 package.json, .npmignore, dist 디렉토리 등을 설정하여 패키지의 버전 관리와 배포를 관리했습니다.

## 📸 스크린샷 (선택 사항)

(필요하다면 구현한 내용을 보여주는 스크린샷 첨부)

## 🔗 참고 링크 (선택 사항)

아래의 링크를 참고하여 npm 배포를 진행했습니다.

- [리액트 컴포넌트 NPM 배포하기 (with TypeScript)](https://velog.io/@junghyeonsu/NPM-배포-어렵지-않아요-with-테오-구글-스프린트-4기)

jeongwusi committed:
📦 chore: 환경 설정

jeongwusi committed:
✨ feat: Container 컴포넌트 구현

jeongwusi committed:
✨ feat: Flex 컴포넌트 구현

jeongwusi committed:
✨ feat: Grid 컴포넌트 구현

jeongwusi committed:
📝 docs: 사용자를 위한 README.md 작성

jeongwusi committed:
📝 docs: 라이센스 부분의 문단 구조 수정

jeongwusi committed:
📦 chore: 버전 수정

jeongwusi committed:
📦 chore: 불필요한 의존성 제거

jeongwusi committed:
📦 chore: .npmignore 파일에 `.github/` 추가

jeongwusi committed:
📦 chore: .gitignore 파일에 /dist 추가와 tsconfig.json 설정 수정

jeongwusi committed:
📦 chore: 라이브러리 버전 0.0.5로 수정

jeongwusi committed:
📦 chore: .npmignore 파일 수정

jeongwusi committed:
♻️ refactor: 파일 구조 수정

jeongwusi committed:
📦 chore: 라이브러리 버전 수정

jeongwusi committed:
📝 docs: Grid 컴포넌트와 Flex 컴포넌트의 Props 관련 내용 수정

jeongwusi committed:
♻️ refactor: gap 속성을 선택적 Props로 수정

jeongwusi committed:
📦 chore: index.ts의 불필요한 주석 제거와 라이브러리 버전 수정

ashleysyheo commented on src/Container/Container.tsx:
```diff
@@ -0,0 +1,17 @@
+
+const Container = ({ children, minWidth, maxWidth }: ContainerProps) => {
+  const containerStyle: React.CSSProperties = {
+    minWidth: `${minWidth}px`,
```
단위를 `px`로 지정해 주셨는데, 이 컴포넌트를 사용하는 사람이 `rem`, `%`, `vw`과 같은 단위를 사용하고 싶을 수도 있지 않을까요?

jeongwusi replied:
넷넷! 사용자가 다양한 단위를 사용할 수 있도록 수정하였으며, 단위는 리터럴 타입을 활용하여 제한을 하였습니다.

```js
interface ContainerProps {
  children: ReactNode;
  minWidth: `${number}${Unit}`;
  maxWidth: `${number}${Unit}`;
.
.
.
```

ashleysyheo commented on src/Container/Container.tsx:
```diff
@@ -0,0 +1,17 @@
+  const containerStyle: React.CSSProperties = {
+    minWidth: `${minWidth}px`,
+    maxWidth: `${maxWidth}px`,
+    margin: "0 auto",
```
`margin` 값을 `"0 auto"`로 지정해준 이유가 있나요? 사용자가 다른 `margin`을 주고 싶을 수 있지 않을까요?

jeongwusi replied:
저는 기본적으로 중앙값을 사용한다고 가정하고 또한 모든 브라우저에 적용할 수 있도록 margin 값을 `"0 auto"`을 주게 되었습니다. 다만 애슐리의 리뷰를 보니 제가 편협한 생각을 가졌다고 생각을 하였습니다. 따라서 margin을 지우고 사용자가 원하는 CSS 스타일을 직접 지정할 수 있도록 수정을 하였습니다.

```js
const containerStyle: CSSProperties = {
    minWidth,
    maxWidth,
    ...style,
  };
.
.
.
```

ashleysyheo commented on src/Container/Container.tsx:
```diff
@@ -0,0 +1,17 @@
+  maxWidth: number;
+}
+
+const Container = ({ children, minWidth, maxWidth }: ContainerProps) => {
```
지금 `Container` 컴포넌트에는 css 관련해서 오로지 `minWidth`, `maxWidth`만 넣을 수 있는데, 사용자가 `backgroundColor`나 `padding` 같은 다른 css 속성을 추가하고 싶을 때는 어떻게 해야 하나요?

jeongwusi replied:
넷넷! 사용자가 CSS 스타일을 직접 지정할 수 있도록 Container 컴포넌트에 style 프로퍼티를 추가하였습니다.

```js
const containerStyle: CSSProperties = {
    minWidth,
    maxWidth,
    ...style,
  };
```

ashleysyheo commented on src/Container/Container.tsx:
```diff
@@ -0,0 +1,17 @@
+    margin: "0 auto",
+  };
+
+  return <div style={containerStyle}>{children}</div>;
```
지금 `div` 컴포넌트만 사용하도록 설정해 줬는데, 사용자가 시맨틱 태그를 사용하고 싶을 때는 어떻게 하나요?

jeongwusi replied:
사용자가 Container 컴포넌트에서 시맨틱 태그를 사용하고 싶을 때를 고려하여, Container 컴포넌트를 수정하여 다른 태그를 사용할 수 있도록 만들었습니다. 이를 위해 as 프로퍼티를 추가하였으며, 원하는 HTML 요소를 렌더링할 수 있도록 할 수 있습니다.

```js
interface ContainerProps {
  children: ReactNode;
  minWidth: `${number}${Unit}`;
  maxWidth: `${number}${Unit}`;
  style?: CSSProperties;
  as?: keyof JSX.IntrinsicElements;
}

const Container = ({
  children,
  minWidth,
  maxWidth,
  style,
  as: Element = "div",
.
.
.
```

ashleysyheo commented on src/Grid/Grid.tsx:
```diff
@@ -0,0 +1,19 @@
+const Grid = ({ children, rows, columns, gap = 0 }: GridProps) => {
+  const gridStyle: React.CSSProperties = {
+    display: "grid",
+    gridTemplateRows: `repeat(${rows}, 1fr)`,
```
지금 `rows`값으로 숫자만 넣을 수 있도록 했는데, 만약 조금 더 복잡한 그리드를 만들고 싶은 경우, `'[row1-start] auto [row2-start] auto [row2-end]'` 이런 식으로 값을 넣을 수도 있지 않을까요?

jeongwusi replied:
더 복잡한 그리드 레이아웃을 만들고 싶을 때, 사용자가 그리드 템플릿을 직접 정의할 수 있도록 Grid 컴포넌트를 개선하였습니다. 이를 위해 gridTemplateRows 및 gridTemplateColumns에 사용자 정의 그리드 템플릿 문자열을 전달할 수 있는 옵션을 추가하였습니다.

```js
.
.
const gridStyle: CSSProperties = {
    display: "grid",
    gridTemplateRows: rows,
    gridTemplateColumns: columns,
.
.
```

ashleysyheo replied:
기존에 했던 것 처럼 숫자를 입력하면 `repeat(${rows}, 1fr)` 이렇게 자동으로 만들어주고 문자열을 입력했을 때만 `gridTemplateRows: rows`, `gridTemplateColumns: columns`, 그 입력 값 자체를 보여줘도 좋을 것 같아요!

ashleysyheo commented on src/index.tsx:
```diff
@@ -0,0 +1,12 @@
+  <React.StrictMode>
+    <App />
+  </React.StrictMode>
+);
```
필요없는 파일은 삭제해도 될 것 같아요!

jeongwusi replied:
넹넹

ashleysyheo commented on package.json:
```diff
@@ -0,0 +1,56 @@
+{
+  "name": "goni-layout-component",
+  "version": "0.0.8",
```
0.0.8 👀

jeongwusi replied:
하하... 시행착오를 많이 겪었습니다 😅

ashleysyheo commented on src/Flex/Flex.tsx:
```diff
@@ -0,0 +1,26 @@
+    | "space-around";
+  align: "flex-start" | "flex-end" | "center" | "stretch";
+  gap?: number;
+}
```
위에 남겼던 질문들과 비슷하게, `flex-wrap`과 같은 속성을 추가하고 싶을 때는 어떻게 해야 하나요??

jeongwusi replied:
넷넷! 해당 컴포넌트 또한 사용자가 CSS 스타일을 직접 지정할 수 있도록 Container 컴포넌트에 style 프로퍼티를 추가하였습니다.

ashleysyheo reviewed (changes requested):
안녕하세요 고니 😊

미션하느라 고생하셨습니다! 궁금한 점 위주로 소소하게 리뷰 남겼습니다. 

#### 질문 

> 컴포넌트 API 설계: 사용자의 입장에서 각 컴포넌트가 어떻게 사용되어야 하는지 고려하여, 필수 및 선택적인 props를 정의했습니다. 또한 TypeScript의 인터페이스를 활용하여 타입 안정성을 보장했습니다.

어떤 기준으로 필수 및 선택적인 props를 정의하셨나요?

<br>

> npm 배포 외에도, 각 컴포넌트의 사용법을 볼 수 있는 페이지를 배포해주세요. 자세한 설명은 README에 작성할 예정이므로, 실제 동작되는 UI 렌더링을 보여준다고 생각해주시면 됩니다. 페이지의 양식은 각 컴포넌트에 있는 [MUI docs](https://mui.com/material-ui/react-container/) 페이지들을 참고해주세요.

LMS에 각 컴포넌트의 사용법을 볼 수 있는 페이지를 배포하라는 요구 사항이 있었는데, 이 부분이 빠진 것 같아요!

jeongwusi committed:
📦 chore: 스토리북 환경 설정

jeongwusi committed:
♻️ refactor: Container 컴포넌트 수정 (관련 내용 상세에 첨부)

- 단위 유연성 개선
- `margin` 프로퍼티 삭제
- 사용자가 필요한 CSS 스타일을 직접 지정할 수 있도록 수정
- 사용자가 시맨틱 태그를 사용하고 싶을 때를 대비하여 `as` 프로퍼티 추가

jeongwusi committed:
♻️ refactor: Flex 컴포넌트 수정 (관련 내용 상세에 첨부)

- 단위 유연성 개선
- 사용자가 필요한 CSS 스타일을 직접 지정할 수 있도록 수정
- 사용자가 시맨틱 태그를 사용하고 싶을 때를 대비하여 `as` 프로퍼티 추가

jeongwusi committed:
♻️ refactor: Grid 컴포넌트 수정 (관련 내용 상세에 첨부)

- 단위 유연성 개선
- 사용자가 필요한 CSS 스타일을 직접 지정할 수 있도록 수정
- 사용자가 시맨틱 태그를 사용하고 싶을 때를 대비하여 `as` 프로퍼티 추가

jeongwusi committed:
✨ feat: 공통으로 사용하는 Unit 타입 추가

jeongwusi committed:
✨ feat: 각 컴포넌트에 맞는 스토리북 코드 작성

jeongwusi committed:
🎨 style: 불필요한 코드 삭제

jeongwusi committed:
📝 docs: README.md 수정

jeongwusi commented:
안녕하세요 애슐리~!! 

애슐리의 친절한 리뷰 덕분에 많은 것을 배울 수 있었습니다. 😊

리뷰에 대한 답변은 코멘트의 댓글과 아래의 내용에 첨부하였습니다!

그럼... 확인 부탁드립니다. 
감사합니다. 🙇🏻 

---

> 어떤 기준으로 필수 및 선택적인 props를 정의하셨나요?

컴포넌트가 제공하는 주요 기능을 고려하여 필수 props를 정의하려고 노력하였습니다. 주어진 미션에서는 사용자의 요구 사항을 보면서 제가 판단하기에 많이 사용될 props는 필수, 아닌 것은 선택으로 나누었습니다.

> LMS에 각 컴포넌트의 사용법을 볼 수 있는 페이지를 배포하라는 요구 사항이 있었는데, 이 부분이 빠진 것 같아요!

아핫... 넹넹 빠진 부분과 수정된 부분 첨부하였습니다.

- 각 컴포넌트의 사용법을 볼 수 있는 [스토리북 페이지](https://nimble-moonbeam-3690a0.netlify.app/?path=/story/container--default)
- 수정된 [npm 배포 페이지](https://www.npmjs.com/package/goni-layout-component)

ashleysyheo commented on src/types/index.ts:
```diff
@@ -0,0 +1 @@
+export type Unit = "px" | "rem" | "%" | "vw";
```
타입을 만들어주셨네요! `vh`, `em` 같은 단위들도 추가적으로 사용할 수 있을 것 같아요!

ashleysyheo commented on src/Container/Container.tsx:
```diff
@@ -0,0 +1,27 @@
+    ...style,
+  };
+
+  return createElement(Element, { style: containerStyle }, children);
```
지금 이렇게 `createElement`를 사용해서 html 요소를 만들어주고 있는데, https://blog.logrocket.com/build-strongly-typed-polymorphic-components-react-typescript/ 이 블로그를 참고하면 JSX를 사용해서 사용자가 입력한 태그로 요소를 만들어 줄 수 있어요!

ashleysyheo commented on src/Container/Container.stories.tsx:
```diff
@@ -0,0 +1,33 @@
+const meta = {
+  title: "Container",
+  component: Container,
+} satisfies Meta<typeof Container>;
```
npm 배포 README.md를 엄청 자세하게 작성해 주셔서 이해하는데 많은 도움이 되었어요! 스토리북에도 그런 설명들이 포함되어 있으면 좋을 것 같아요! 이 라이브러리를 실제로 사용할 사람들이 리듬이 보다는 스토리북을 참고해서 어떻게 사용할지를 살펴볼 것 같은데, 이때 props에 대한 설명과 이 컴포넌트가 뭘 하는지 알 수 있으면 좋을 것 같아요.

<img width="642" alt="Screenshot 2023-09-20 at 11 40 53 PM" src="https://github.com/woowacourse/layout-component/assets/51967731/300613ec-a0cd-4f7a-95a2-67dd5c58a50d">

이런식으로 jsdoc 주석을 추가해주면 스토리북에서도 각 props에 대한 설명을 볼 수 있어요! 그리고 https://storybook.js.org/docs/react/writing-docs/mdx 이 링크를 참고하면 스토리북 문서도 더 풍부하게(?) 만들 수 있을 것 같아요 😊

ashleysyheo reviewed (approved):
안녕하세요 고니! 

안녕하세요 고니! 

리뷰 반영한 거 잘 확인했습니다!! 추가적으로 고민해보면 좋을 것 같은 부분들에 코멘트 남겨봤습니다 😊

미션하느라 수고하셨습니다 👍🔥💯 step에서 뵙겠습니다!
